aws/resource_aws_sns_platform_application.go

```diff
@@ -0,0 +1,272 @@
+package aws
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/service/sns"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+var snsPlatformRequiresPlatformPrincipal = map[string]bool{
+	"APNS":         true,
+	"APNS_SANDBOX": true,
+}
+
+// Mutable attributes
+// http://docs.aws.amazon.com/sns/latest/api/API_SetPlatformApplicationAttributes.html
+var snsPlatformApplicationAttributeMap = map[string]string{
+	"event_delivery_failure_topic_arn": "EventDeliveryFailure",
+	"event_endpoint_created_topic_arn": "EventEndpointCreated",
+	"event_endpoint_deleted_topic_arn": "EventEndpointDeleted",
+	"event_endpoint_updated_topic_arn": "EventEndpointUpdated",
+	"failure_feedback_role_arn":        "FailureFeedbackRoleArn",
+	"platform_principal":               "PlatformPrincipal",
+	"success_feedback_role_arn":        "SuccessFeedbackRoleArn",
+	"success_feedback_sample_rate":     "SuccessFeedbackSampleRate",
+}
+
+func resourceAwsSnsPlatformApplication() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsSnsPlatformApplicationCreate,
+		Read:   resourceAwsSnsPlatformApplicationRead,
+		Update: resourceAwsSnsPlatformApplicationUpdate,
+		Delete: resourceAwsSnsPlatformApplicationDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		CustomizeDiff: func(diff *schema.ResourceDiff, v interface{}) error {
+			return validateAwsSnsPlatformApplication(diff)
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"platform": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"platform_credential": {
+				Type:      schema.TypeString,
+				Required:  true,
+				StateFunc: hashSum,
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"event_delivery_failure_topic_arn": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"event_endpoint_created_topic_arn": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"event_endpoint_deleted_topic_arn": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"event_endpoint_updated_topic_arn": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"failure_feedback_role_arn": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"platform_principal": {
+				Type:      schema.TypeString,
+				Optional:  true,
+				StateFunc: hashSum,
+			},
+			"success_feedback_role_arn": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"success_feedback_sample_rate": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
```

Comment:
I'm not sure, honestly, 🙁 . The precedent here is that the other SNS resources (topic/topic_subscription) don't nest them. In this case, I think we're okay because the AWS API nesting is really just an implementation detail and its slightly easier Terraform code and user experience to keep these as top level attributes. Of course that's not valid if we were planning on supporting something like a `aws_sns_platform_application_attribute` resource (and do the same to the others).

If you feel strongly I can definitely change it. 👍

Reply:
It really depends how is the API going to evolve going forward - i.e. whether there is going to be more arguments on the root level (which could potentially clash with any nested ones).

I think it's a guessing game at this point and we have quite good deprecation mechanisms available in the schema, so it's fine to keep it as is for now.